idea-plugin/src/main/kotlin/arrow/meta/plugin/idea/plugins/higherkinds/HigherKindsIdePlugin.kt

```diff
@@ -1,20 +1,124 @@
+import org.jetbrains.kotlin.psi.KtParameter
 import org.jetbrains.kotlin.utils.addToStdlib.safeAs
 
 val IdeMetaPlugin.higherKindsIdePlugin: Plugin
   get() = "HigherKindsIdePlugin" {
     meta(
-      addLineMarkerProvider(
-        icon = ArrowIcons.HKT,
-        message = "HKT",
-        matchOn = { it.safeAs<KtClass>()?.let(::isHigherKindedType) == true }
+      lineMarker(
+        matchOn = { it.safeAs<KtClass>()?.let(::isHigherKindedType) == true },
+        marker = {
+          LineMarker(
+            icon = ArrowIcons.HKT,
+            message = it.safeAs<KtClass>()?.let { classOrInterface ->
+              """|${classOrInterface.name} is a Higher Kinded Type that may be used in polymorphic functions expressed over [Kind<F, A>] and with the type classes. 
+                 |For more info visit [https://arrow-kt.io/docs/patterns/glossary/#type-constructors](https://arrow-kt.io/docs/patterns/glossary/#type-constructors)
+                 |""".trimMargin()
+            } ?: ""
+          )
+        }
+      ),
+      lineMarker(
+        matchOn = { it.safeAs<KtNamedFunction>()?.run { isKindPolymorphic() && hasExtensionDefaultValues() } == true },
+        marker = { element ->
+          LineMarker(
+            icon = ArrowIcons.POLY,
+            message = element.safeAs<KtNamedFunction>()?.let { f ->
+              FuncScope(f).run {
+                val target = ScopedList(receiver.value, transform = { it.text.escapeHTML()})
+                """|<h1>$name</h1>
+                   |<code lang="kotlin">
+                   |  ${this@run} 
+                   |</code>
+                   |<code>$name</code> is a polymorphic function that can be applied over $target.
+                   |<code>$name</code> requires @extension proofs for <code>${givenParameters()}</code>
+                   |
+                   |Valid <code>@extension</code> locations:
+                   |${validExtensionLocations()}
+                   |
+                   |<a href="https://arrow-kt.io/docs/patterns/polymorphic_programs/">Learn more about type class and ad-hoc polymorphism</a>
+                   |""".trimMargin()
+              }
+            } ?: ""
+          )
+        }
       )
     )
   }
+
+private fun FuncScope.givenParameters(): ScopedList<KtParameter> =
+  ScopedList(`(valueParameters)`.value.filter { it.hasExtensionDefaultValue() }, transform = {
+    it.text.escapeHTML()
+  })
+
+private fun FuncScope.validExtensionLocations(): String =
+  givenParameters().value.joinToString("\n") {
+    val type = it.type()
+    val extensionType = type.toString().escapeHTML()
+    val typeClassDescriptor = type?.constructor?.declarationDescriptor
+    val typeClassName = typeClassDescriptor?.name?.asString()
+    val typeClassPackage = typeClassDescriptor?.importableFqName?.pathSegments()?.dropLast(1)?.joinToString(".")
+    val localPackage = value.containingKtFile.packageFqName
+    """|<h2>Type Class: $typeClassPackage</h2>
```

Comment:
I'm a little confused about this message. Are we going to point to the resolved instance later on?
Or is this meant to educate about typeclasses?

Reply:
Resolved: just saw first-pass in branch name ;)